tests/i18n/tests.py

```diff
@@ -384,6 +384,50 @@ def test_language_bidi_null(self):
             self.assertIs(get_language_bidi(), True)
 
 
+class TranslationLoadingTests(SimpleTestCase):
+
+    def setUp(self):
+        """Clear translation state."""
+        self._old_language = get_language()
+        self._old_translations = trans_real._translations
+        deactivate()
+        trans_real._translations = {}
+
+    def tearDown(self):
+        trans_real._translations = self._old_translations
+        activate(self._old_language)
+
+    @override_settings(
+        USE_I18N=True,
+        LANGUAGE_CODE='en',
+        LANGUAGES=[
+            ('en', 'English'),
+            ('en-ca', 'Canadian'),
+            ('en-nz', 'New Zealandian'),
+            ('en-au', 'Australian'),
+        ],
+        LOCALE_PATHS=[os.path.join(here, 'loading')],
+        INSTALLED_APPS=['i18n.loading_app'],
+    )
+    def test_translation_loading(self):
+        """
+        ``loading_app`` does not have translations for all languages provided
+        by ``loading``. Catalogs must be merged correctly.
+        """
+        tests = [
+            ('en', 'local country person'),
+            ('en_AU', 'aussie'),
+            ('en_NZ', 'kiwi'),
+            ('en_CA', 'canuck'),
+        ]
+        # All translations are loaded by the second pass.
+        for rnd in range(1, 3):
+            for language, nickname in tests:
+                with self.subTest(language=language, round=rnd):
+                    activate(language)
+                    self.assertEqual(gettext('local country person'), nickname)
```

Comment:
I think we can remove `tearDown()` and `setUp()` and use
```python
with translation.override(language):
```
instead of `activate()`.

Reply:
I think we do need this. 

We need to reset the translations totally so that they are loaded as we activate each language. We could use override but then we're mixing in with the translations that have already been setup.

Reply:
OK, so maybe we can at remove `active()/deactivate()` round-trip with:
```diff
diff --git a/tests/i18n/tests.py b/tests/i18n/tests.py
index ccf4c242cf..c7be1f2634 100644
--- a/tests/i18n/tests.py
+++ b/tests/i18n/tests.py
@@ -390,12 +390,10 @@ class TranslationLoadingTests(SimpleTestCase):
         """Clear translation state."""
         self._old_language = get_language()
         self._old_translations = trans_real._translations
-        deactivate()
         trans_real._translations = {}
 
     def tearDown(self):
         trans_real._translations = self._old_translations
-        activate(self._old_language)
 
     @override_settings(
         USE_I18N=True,
@@ -423,8 +421,7 @@ class TranslationLoadingTests(SimpleTestCase):
         # All translations are loaded by the second pass.
         for rnd in range(1, 3):
             for language, nickname in tests:
-                with self.subTest(language=language, round=rnd):
-                    activate(language)
+                with self.subTest(language=language, round=rnd), translation.override(language):
                     self.assertEqual(gettext('local country person'), nickname)
 
```
It works the same for me.

Reply:
Yeah... 

I guess that's OK. I'm not as keen, because the `activate()` call is what's doing the work here: it's loading the translation. 

Then we have it double context manager line, rather than in the straight-line of the test code. For me, it's kind-of hiding the code under test. 

I don't really mind, but that's why I wrote it the way I did. 🤷

Reply:
I don't think we need to reset those here (although it wouldn't hurt) 

They [map the underlying GNUTranslations to mo files](https://github.com/python/cpython/blob/a63c61168588937c482435d0432c753de4844c46/Lib/gettext.py#L575-L576). The issue we have here is how we merge those into DjangoTranslations — so we need to reset our dict, in `trans_real._translations`. (**I think** 🙂)

It looks like we can get away without doing that with the current tests, but I think there'd be a independence issue...